### PR TITLE
fix: [CDS-46922]: bug fixes for service/env at o/a level

### PR DESCRIPTION
**Pull request by hingerabhinav**
## Describe your changes
1. throw exception with correct yaml field for entity validation in service,env
2. build service step outcome with scoped ref
3. build environment outcome with scoped ref.
4. made params non mandatory for fetching runtime inputs 
5. use scope based on service ref for validating access in Service step
6. modify infra request so that it is always the same scope as the environment, regardless if other params are passed (for inline creation of infra in pipeline studio)

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/42459)

### CODE REVIEW

**Comment by tchaurasiya** [comment]
*[`127-cd-nextgen-entities/src/main/java/io/harness/cdng/service/steps/ServiceStepOutcome.java` line 65 at r3](https://reviewable.io/reviews/harness/harness-core/42459#-NLUwpXK3rSJGOz9vZJL:-NLUwpXK3rSJGOz9vZJM:b-549amz) ([raw file](https://github.com/harness/harness-core/blob/fab01fd3331f4cd828bc9361a21aa0a4e36e463f/127-cd-nextgen-entities/src/main/java/io/harness/cdng/service/steps/ServiceStepOutcome.java#L65)):*
> ```Java
>   }
> 
>   public static ServiceStepOutcome fromServiceStepV2(String identifier, String name, String type, String description,
> ```

nit: Lets change parameter name identifier to identifierRef
to aware caller of this method that ServiceOutcome needs qualifiedIdentifierRef.

**Comment by tchaurasiya** [comment]
*[`127-cd-nextgen-entities/src/main/java/io/harness/ng/core/infrastructure/services/impl/InfrastructureEntityServiceImpl.java` line 661 at r3](https://reviewable.io/reviews/harness/harness-core/42459#-NLUxqhr627lMENQvlGJ:-NLUxqhr627lMENQvlGK:bqcqd89) ([raw file](https://github.com/harness/harness-core/blob/fab01fd3331f4cd828bc9361a21aa0a4e36e463f/127-cd-nextgen-entities/src/main/java/io/harness/ng/core/infrastructure/services/impl/InfrastructureEntityServiceImpl.java#L661)):*
> ```Java
>     requestInfra.setName(requestInfra.getName().trim());
>     // convert to scope of the environment
>     String[] envRefSplit = StringUtils.split(requestInfra.getEnvIdentifier(), ".", MAX_RESULT_THRESHOLD_FOR_SPLIT);
> ```

nit: lets use your method itself io.harness.utils.FullyQualifiedIdentifierHelper#getRefFromIdentifierOrRef

**Review by tchaurasiya — DISMISSED** [review]
> Reviewed 7 of 10 files at r1, 6 of 7 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42459)* status: 14 file(s) to review. 1 of 2 approvals obtained.

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepV3.java [hunk 1]
```diff
@@ -463,7 +464,7 @@ public StepResponse handleChildrenResponse(
 
     final ServiceSweepingOutput serviceSweepingOutput = (ServiceSweepingOutput) sweepingOutputService.resolve(
         ambiance, RefObjectUtils.getOutcomeRefObject(ServiceStepV3.SERVICE_SWEEPING_OUTPUT));
-
+    String serviceRef = stepParameters.getServiceRef().getValue().trim();
```
Comment (yogesh-chauhan) [inline 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepV3.java:467]: is there a reason we need to trim this string ? Lets use it as is.

**Review by yogesh-chauhan — CHANGES_REQUESTED** [review]

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepV3.java (hunk 1 above)
Comment (hingerabhinav) [reply 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepV3.java:467]: Ack

**Review by tchaurasiya — APPROVED** [review]
> Reviewed 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42459)* status: 14 file(s) to review. 1 change requests. 1. unresolved discussions. 1 of 2 approvals obtained.

**Review by yogesh-chauhan — APPROVED** [review]
> Reviewed 1 of 10 files at r1, 2 of 7 files at r2, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42459)* status: :shipit: complete! ✓

**Pull request merged**